lib/SILOptimizer/Transforms/ArrayElementValuePropagation.cpp

```diff
@@ -249,37 +304,75 @@ bool ArrayAllocation::collectForwardableValues() {
 namespace {
 
 class ArrayElementPropagation : public SILFunctionTransform {
+
 public:
   ArrayElementPropagation() {}
 
   StringRef getName() override {
     return "Array Element Propagation";
   }
 
+  void replaceAppendCalls(SmallVector<ValueReplacementsPair, 4> Repls) {
+    auto &Fn = *getFunction();
+    auto &M = Fn.getModule();
+    auto &C = M.getASTContext();
+
+    if (Repls.empty())
+      return;
+
+    llvm::dbgs() << "Array append contentsOf calls replaced in "
+                 << Fn.getName() << " ("
+                 << Repls.size() << ")\n";
+
+    auto *AppendFnDecl = C.lookupArrayAppendElementDecl();
+    if (!AppendFnDecl)
+      return;
+
+    auto Mangled = SILDeclRef(AppendFnDecl, SILDeclRef::Kind::Func).mangle();
+    auto *AppendFn = M.hasFunction(Mangled, SILLinkage::PublicExternal);
+    if (!AppendFn)
+      return;
+
+    for (auto &Repl : Repls) {
+      ArraySemanticsCall AppendContentsOf(Repl.first);
+      assert(AppendContentsOf && "Must be AppendContentsOf call");
+      assert(Repl.second.size() && "Must have at least one replacement");
+      auto Ty = Repl.second[0]->getType().getSwiftType();
+      ArrayRef<Substitution> Subs{Substitution(Ty, {})};
```

Comment:
I'm still a bit nervous about using a lowered type here. Can you check what happens if you perform your optimization with a substitution where T maps to a function type?